ui/dashboards/src/components/Variables/TemplateVariable.tsx

```diff
@@ -74,29 +82,68 @@ function ListVariable({ name, source }: TemplateVariableProps) {
   const valueIsInOptions = useMemo(
     () =>
       Boolean(
-        finalOptions.find((v) => {
+        viewOptions.find((v) => {
           if (allowMultiple) {
             return (value as string[]).includes(v.value);
           }
           return value === v.value;
         })
       ),
-    [finalOptions, value, allowMultiple]
+    [viewOptions, value, allowMultiple]
   );
 
-  let selectValue = value;
-  if (!valueIsInOptions) {
-    selectValue = allowMultiple ? [] : '';
-  }
+  value = useMemo(() => {
+    const firstOptionValue = viewOptions?.[allowAllValue ? 1 : 0]?.value;
+
+    // If there is no value but there are options, or the value is not in options, we set the value to the first option.
+    if (firstOptionValue) {
+      if (!valueIsInOptions || !value || value.length === 0) {
+        return allowMultiple ? [firstOptionValue] : firstOptionValue;
+      }
+    }
+
+    return value;
+  }, [viewOptions, value, valueIsInOptions, allowMultiple, allowAllValue]);
+
+  // Once we computed value, we set it as the selected one, if it is available in the options
+  const selectedValue = value && valueIsInOptions ? value : allowMultiple ? [] : '';
 
+  return { value, loading, options, selectedValue, viewOptions };
+}
+
+function ListVariable({ name, source }: TemplateVariableProps) {
+  const ctx = useTemplateVariable(name, source);
+  const definition = ctx.definition as ListVariableDefinition;
+  const variablesOptionsQuery = useListVariablePluginValues(definition);
+  const { setVariableValue, setVariableLoading, setVariableOptions } = useTemplateVariableActions();
+  const { selectedValue, value, loading, options, viewOptions } = useListVariableState(
+    definition?.spec,
+    ctx.state,
+    variablesOptionsQuery
+  );
+
+  const title = definition?.spec.display?.name ?? name;
+  const allowMultiple = definition?.spec.allow_multiple === true;
+  const allowAllValue = definition?.spec.allow_all_value === true;
+
+  // Update value when changed
   useEffect(() => {
-    const firstOption = finalOptions?.[0];
+    if (value) {
+      setVariableValue(name, value, source);
+    }
+  }, [setVariableValue, name, value, source]);
+
+  // Update loading when changed
```

Comment:
Approved but I have a hunch that some of these hooks could be refactored and removed, didn’t get a chance to look into more specifics but newer React docs are emphasizing that useEffect can be unnecessary in many cases https://react.dev/learn/you-might-not-need-an-effect

Reply:
I fully agree to refactor it! Here I just encapsulate and test but for sure it doesn't make the code more readable, just a bit safer. Also in this particular case, it's tricky as we don´t want to update value too often or we have an infinite loop with the subsequent setVariableValue :smile:

Reply:
I can try to remove the useEffect for loading and options, but I'm sure that removing it for value would do some infinite loop...